src/topology/builder.rs

```diff
@@ -562,3 +412,242 @@ const fn filter_event_type(event: &Event, data_type: DataType) -> bool {
+        // TODO: account for named outputs separately?
+        let count = outputs_buf.len();
+        // TODO: do we only want allocated_bytes for events themselves?
+        let byte_size = outputs_buf.size_of();
+
+        self.timer.start_wait();
+        self.outputs.send(outputs_buf).await;
+
+        emit!(&EventsSent { count, byte_size });
+    }
+
+    async fn run_inline(
+        mut self,
+        mut transform: Box<dyn SyncTransform>,
+        input_rx: BufferStream<Event>,
+        input_type: DataType,
+    ) -> Result<TaskOutput, ()> {
+        let mut outputs_buf = self.outputs.new_buf_with_capacity(TRANSFORM_BATCH_SIZE);
+
+        let mut input_rx = input_rx
+            .filter(move |event| ready(filter_event_type(event, input_type)))
+            .ready_chunks(TRANSFORM_BATCH_SIZE);
+
+        self.timer.start_wait();
+        while let Some(events) = input_rx.next().await {
+            self.on_events_received(&events);
+
+            for event in events {
+                transform.transform(event, &mut outputs_buf);
+            }
+
+            self.send_outputs(&mut outputs_buf).await;
+        }
+
+        debug!("Finished.");
+        Ok(TaskOutput::Transform)
+    }
+
+    async fn run_concurrently(
+        mut self,
+        transform: Box<dyn SyncTransform>,
+        input_rx: BufferStream<Event>,
+        input_type: DataType,
+    ) -> Result<TaskOutput, ()> {
+        let mut input_rx = input_rx
+            .filter(move |event| ready(filter_event_type(event, input_type)))
+            .ready_chunks(TRANSFORM_BATCH_SIZE);
+
+        let mut in_flight = FuturesOrdered::new();
+        let mut shutting_down = false;
+
+        self.timer.start_wait();
+        loop {
+            tokio::select! {
+                biased;
+
+                result = in_flight.next(), if !in_flight.is_empty() => {
+                    match result {
+                        Some(Ok(outputs_buf)) => {
+                            let mut outputs_buf: TransformOutputsBuf = outputs_buf;
+                            self.send_outputs(&mut outputs_buf).await;
+                        }
+                        _ => unreachable!("join error or bad poll"),
+                    }
+                }
+
+                input_events = input_rx.next(), if in_flight.len() < CONCURRENCY_LIMIT && !shutting_down => {
+                    match input_events {
+                        Some(events) => {
+                            self.on_events_received(&events);
+
+                            let mut t = transform.clone();
+                            let mut outputs_buf = self.outputs.new_buf_with_capacity(events.len());
+                            let task = tokio::spawn(async move {
+                                for event in events {
+                                    t.transform(event, &mut outputs_buf);
+                                }
+
+                                outputs_buf
+                            });
```

Comment:
If I am reading right, each chunk of the input is spun into a separate task, which may be put onto separate threads. So, if a single client connects to a source with acknowledgements enabled, it will never have any work done in parallel. Has the use of something like [rayon](https://docs.rs/rayon/latest/rayon/) been discussed, where each batch would be split across threads?

Reply:
> So, if a single client connects to a source with acknowledgements enabled, it will never have any work done in parallel.

I'm missing a step here. Why would acknowledgements being enabled mean no work would be done in parallel?

> Has the use of something like rayon been discussed, where each batch would be split across threads?

Rayon is an excellent library but it's for different things. Rayon runs its own fixed-size threadpool in competition with the dynamic pool of threads that tokio maintains for its scheduler. It _is_ possible to have a setup where tokio handles IO solely and we kick CPU-bound payloads into another task that owns a rayon threadpool via a channel, then await for the result on another channel but I don't think that pattern of work fits here.

Reply:
I was thinking that if acknowledgements are enabled, the source would have to wait for acks before issuing another batch, hence only one would be in flight at any time. Oh, unless the batch was larger than the chunk size. This actually means having a *smaller* chunk size would improve performance, contrary to what I said above about the size. As I type this, I am realizing that pipelining within the protocol could result in parallel batches too. So, not quite what I was describing.

Yeah, I was thinking using rayon directly would not work well, hence the "something like" statement. ISTR discussion that follow-up work would look into tuning this chunk scheduling to see if another algorithm works better, so I am fine to leave this as-is given the current speed up.

Reply:
> I was thinking that if acknowledgements are enabled, the source would have to wait for acks before issuing another batch, hence only one would be in flight at any time.

I believe this will be a more general performance challenge with acknowledgements, and not necessarily specific to this concurrency work. We could try to do something like reduce the chunk size, but that feels iffy. The real answer is likely to find some way to pipeline batches from the source. How feasible that is will likely depend heavily on the source protocol, but it's probably the first avenue to explore.